README
SLURM BANK
==========

SLURM Bank, a collection of wrapper scripts to give slurm GOLD like
capabilities for managing resources.

With the scripts we are able to provide a simple banking system where we
can deposit hours to an account. Users are associated to these accounts
from which they can use to run jobs. If users do not have an account or
if they do not have hours in their account then they cannot run jobs.

Requirements (tested with)

* SLURM 2.2.0, 2.2.7, 2.4.1, 2.5.3
* Scientificlinux 5.x (bash, rsync, perl)

Source and Documentation
========================

The documentation is stored in doc/ and is generated with ikiwiki,
that is, all the documentation is written in markdown.

There is a pre-generated set of documentation available at
<http://jcftang.github.com/slurm-bank/>

To get the source and documentation please do

	git clone git://github.com/jcftang/slurm-bank.git

Administrator Notes
===================

To build an installable RPM from the tarball

	rpmbuild -ta slurm-bank-1.1.1.2.tar.gz

The above assumes that the appropriate compiler and rpm-build packages
are installed on your system. You will need ikiwiki installed to generate
the html documetation. If you do not have ikiwiki installed you can do:

	rpmbuild -ta --without docs slurm-bank-1.1.1.2.tar.gz

If you have a clone of the repository, there is a pre-generated set of
html pages stored in the html branch.

Tests
=====

There is a simple test suite to make sure things are working as we
expect. To execute them,

	make test

The tests behave differently if you run them as a normal user or as root.

Doing a release
===============

* clone the git repository
* Finalise changes
* Run bump-version script to make some version number changes to various files
* Tag repository
* then run "make dist" to generate a tarball for redistribution


ChangeLog
2013-06-06  Paddy Doyle <paddy@tchpc.tcd.ie>
	* Version: 1.3.2
	* src/_sbank-balance.pl: Add a '-s yyyy-mm-dd' start-date parameter
	  to balance report.

2013-05-30  Paddy Doyle <paddy@tchpc.tcd.ie>
	* Version: 1.3.1
	* src/_sbank-balance.pl: Change the way usage data is pulled out
	  of the system: change from 'sshare' to 'sreport', which means
	  pulling from the Slurmdbd instead of local *_usage files.

2012-10-05  Jimmy Tang <jtang@tchpc.tcd.ie>
	* Version: 1.2.1


t/test.sh
#!/usr/bin/env bash

. ../wvtest.sh

#set -e

# cluster, account and user
_cluster=tcluster
_account=taccount
_user=tuser
_root=`id -g`

sbank()
{
    "../src/sbank" "$@"
}

WVSTART "init"

WVPASS which perl
WVPASS which scontrol
WVPASS which sacctmgr
WVPASS which sshare
WVPASS which sinfo
WVPASS which sbatch

WVFAIL sbank

WVSTART "sbank time"

WVPASSEQ "$(sbank time calc -t 4-00:00:00)" "96"
WVPASSEQ "$(sbank time estimate -N 4 -c 8 -t 24)" "768"
WVPASSEQ "$(sbank time estimate -n 32 -t 96)" "3072"
WVPASSEQ "$(sbank time estimate -n 32 -t $(sbank time calc -t 4-00:00:00))" "3072"
WVPASSEQ "$(sbank time estimate -n 64 -t 96)" "$(sbank time estimate -n 64 -t $(sbank time calc -t 4-00:00:00))"

WVSTART "sbank balance"

WVPASS sbank time estimatescript -s sample-job1.sh
WVPASS sbank time estimatescript -s sample-job2.sh


if [ "${_root}" -eq "0" ];
then
WVSTART "sbank cluster - bits that need root"
WVPASS sbank cluster create -c $_cluster
WVPASS sbank cluster delete -c $_cluster
else
WVSTART "skip sbank cluster (create/delete) - root required"
fi

WVSTART "sbank cluster - bits that don't need root"
WVPASSRC sbank cluster cpupernode
WVPASSRC sbank cluster cpupernode -m
WVPASSRC sbank cluster list
WVPASSRC sbank cluster list -a

if [ "${_root}" -eq "0" ];
then
WVSTART "sbank project - bits that need root"

WVPASS sbank cluster create -c $_cluster
WVPASS sbank project create -c $_cluster -a $_account
WVPASSEQ "$(sbank project list -c $_cluster | awk '{print $2}' | grep $_account)" "$_account"
WVPASS sbank project delete -c $_cluster -a $_account
WVPASS sbank cluster delete -c $_cluster
else
WVSTART "skip sbank project - root required"
fi

if [ "${_root}" -eq "0" ];
then
WVSTART "sbank deposit - bits that need root"

WVPASS sbank cluster create -c $_cluster
WVPASS sbank project create -c $_cluster -a $_account
WVPASSEQ "$(sbank project list -c $_cluster | awk '{print $2}' | grep $_account)" "$_account"
WVPASS sbank deposit -c $_cluster -a $_account -t 1000
WVPASSEQ "$(sbank balance request -a taccount -c tcluster -t 500)" "-500"
WVPASS sbank project delete -c $_cluster -a $_account
WVPASS sbank cluster delete -c $_cluster
else
WVSTART "skip sbank deposit - root required"
fi


src/sbank-balance.pl
#!/usr/bin/perl -w

# Emulate Gold's 'mybalance' script for SLURM
# 
# Assumes Enforced Limits, with GrpCPUMins set on Accounts (not users)
# 
# Specifically, the following must be set in slurm.conf:
#     AccountingStorageEnforce=limits
#     PriorityDecayHalfLife=0
#     PriorityUsageResetPeriod=NONE
# 
# Requires 'sacctmgr' and 'sshare'
# 

# TODO:
# - re-write using SLURM-Perl API


use strict;
use Getopt::Std;


my %acc_limits = ();
my %acc_usage = ();
my %user_usage = ();
my $thisuser = (getpwuid($<))[0];	# who's running the script
my $showallusers = 0;
my $showallaccs = 0;
my $clustername = "";
my $accountname = "";
my ($account, $user, $rawusage, $prev_acc);


#####################################################################
# subroutines
#####################################################################
sub usage() {
	print "Usage:\n";
	print "$0 [-h] [-c clustername] [-a accountname] [-A]\n";
	print "\t-h:\tshow this help message\n";
	print "\t-c:\tdisplay per cluster 'clustername' (defaults to the local cluster)\n";
	print "\t-a:\tdisplay unformatted balance of account 'accountname' (defaults to all accounts of the current user)\n";
	die   "\t-A:\tdisplay all accounts (defaults to all accounts of the current user; implies '-u')\n";
}

# format minutes as hours, with thousands comma separator
sub fmt_mins_as_hrs( $ ) {
	my $n = shift;

	return thous(sprintf("%.0f", $n/60));
}

# add commas
sub thous( $ ) {
	my $n = shift;
	1 while ($n =~ s/^(-?\d+)(\d{3})/$1,$2/);
	return $n;
}


#####################################################################
# get options
#####################################################################
my %opts;
getopts('huc:a:A', \%opts) || usage();

if (defined($opts{h})) {
	usage();
}

if (defined($opts{u})) {
	$showallusers = 1;
}

if (defined($opts{c})) {
	$clustername = $opts{c};
}

if (defined($opts{a})) {
	$accountname = $opts{a};
}

if (defined($opts{A})) {
	$showallaccs = 1;
	$showallusers = 1;	# implied
}


#####################################################################
# start
# get the local clustername, or use the given clustername
#####################################################################

if ($clustername eq "") {
	open (SCONTROL, 'scontrol show config |')
		or die "$0: Unable to run scontrol: $!\n";

	while (<SCONTROL>) {
		if (/^ClusterName\s*=\s*(\S+)/) {
			$clustername = $1;
		}
	}

	close(SCONTROL);

	if ($clustername eq "") {
		die "$0: Unable to determine local cluster name via scontrol. Exiting..\n";
	}
}


#####################################################################
# run sacctmgr to find all Account limits from the list of 
# Assocations
#####################################################################

open (SACCTMGR, "sacctmgr list association cluster=$clustername format='Account,GrpCPUMins'" .
		" -p -n |")
	or die "$0: Unable to run sacctmgr: $!\n";

# GrpCPUMins are not in 'sshare'
while (<SACCTMGR>) {
	# format is "acct_string|nnnn|" where nnnn is the number of GrpCPUMins allocated
	if (/^([^|]+)\|([^|]*)/) {
		if ($2 ne "") {
			$acc_limits{"\U$1"} = sprintf("%.0f", $2);
		}
	}
}

close(SACCTMGR);


#####################################################################
# quick sanity check - did we find any GrpCPUMins ?
#####################################################################

if ((scalar keys %acc_limits) == 0) {
	die "$0: Unable to find any GrpCPUMins set on Accounts in cluster '$clustername' via sacctmgr. Exiting..\n";
}


#####################################################################
# if we need to show all users in our Accounts, then we have to
# run sshare twice - once to find all Accounts that I'm a part of,
# and secondly to dump all users in those accounts
#####################################################################

if ($showallusers) {
	# complex version:
	# show all users in only my Accounts

	my @my_accs = ();
	my $cluster_str = ($clustername ne "") ? "-M $clustername " : "";
	my $account_str = ($showallaccs) ? "-a " : "";

	$account = "";	# init to a value to stop perl warnings

	# only look to a specified account, rather than all
	# note we still need to run the loop below, to gather the account usage
	# values for all accounts
	if ($accountname ne "") {
		push @my_accs, $accountname;
	}


	# firstly grab the list of accounts that I'm a part of
	open (SSHARE, "sshare $account_str $cluster_str -hp |") or die "$0: Unable to run sacctmgr: $!\n";

	while (<SSHARE>) {
		# only show outputs for accounts we're part of
		if (/^\s*([^|]+)\|([^|]*)\|[^|]*\|[^|]*\|([^|]*)/) {
			$account      = "\U$1";
			$user         = $2;
			$rawusage     = $3;

			if ($user ne "") {
				# a user field present means we're part of the account, so add it;
				# unless we're only looking for a specific account
				if ($accountname eq "") {
					# build up list of my accounts
					push @my_accs, $account;
				}
			} elsif (exists($acc_limits{$account})) {
				# and store the account limits while we're here
				$acc_usage{$account} = sprintf("%.0f", $rawusage/60);
			}
		}
	}

	close(SSHARE);


	# display formatted output
	printf "%-10s %11s | %16s %11s | %13s %11s (CPU hrs)\n",
		"User", "Usage", "Account", "Usage", "Account Limit", "Available";
	printf "%10s %11s + %16s %11s + %13s %11s\n",
		"-"x10, "-"x11, "-"x16, "-"x11, "-"x13, "-"x11;


	# now, display all users from just those accounts
	open (SSHARE2, "sshare $cluster_str -aA " . (join ",",@my_accs) . " -hp 2>/dev/null |") or die "$0: Unable to run sacctmgr: $!\n";

	$prev_acc = "";

	while (<SSHARE2>) {
		if (/^\s*([^|]+)\|([^|]*)\|[^|]*\|[^|]*\|([^|]*)/) {
			if ($prev_acc ne "" && $account ne $prev_acc) {
				print "\n";
			}
			if ($account ne "") {
				$prev_acc = $account;
			}

			$account      = "\U$1";
			$user         = $2;
			$rawusage     = $3;

			next if($user eq "");

			# highlight current user
			if ($user eq $thisuser) {
				$user = "$user *";
			}

			# stop warnings if this account doesn't have a limit
			if (! exists($acc_limits{$account})) {
				$acc_limits{$account} = 0;
			}

			# stop warnings if this account doesn't have any usage
			if (! exists($acc_usage{$account})) {
				$acc_usage{$account} = 0;
			}

			printf "%-10s %11s | %16s %11s | %13s %11s\n",
				$user, fmt_mins_as_hrs(sprintf("%.0f", $rawusage/60)),
				$account, fmt_mins_as_hrs($acc_usage{$account}),
				fmt_mins_as_hrs($acc_limits{$account}),
				fmt_mins_as_hrs($acc_limits{$account} - $acc_usage{$account});
		}
	}

	close(SSHARE2);

} elsif ($accountname ne "") {
	# show only the balance for $accountname, unformatted

	my $cluster_str = ($clustername ne "") ? "-M $clustername " : "";

	$rawusage = "";	# init to a value to stop perl warnings

	# grab the usage for just that account
	open (SSHARE, "sshare $cluster_str -A $accountname -hp 2>/dev/null |") or die "$0: Unable to run sacctmgr: $!\n";

	# the first line of input has the values
	$_ = <SSHARE>;
	if (/^\s*([^|]+)\|([^|]*)\|[^|]*\|[^|]*\|([^|]*)/) {
		$account      = "\U$1";
		$user         = $2;
		$rawusage     = $3;

		if (exists($acc_limits{$account})) {
			# and store the account limits while we're here
			$acc_usage{$account} = sprintf("%.0f", $rawusage/60);
		}
	}

	close(SSHARE);

	if ($rawusage eq "") {
		die "$0: invalid account string '$accountname'\n";
	}

	# this is minutes - we need to convert to hours
	printf "%.0f\n", (($acc_limits{$account} - $acc_usage{$account})/60);

} else {
	# simple version:
	# only show my usage in the Accounts

	my $cluster_str = ($clustername ne "") ? "-M $clustername " : "";

	open (SSHARE, "sshare $cluster_str -hp |") or die "$0: Unable to run sacctmgr: $!\n";

	while (<SSHARE>) {
		# only show outputs for accounts we're part of
		if (/^\s*([^|]+)\|([^|]*)\|[^|]*\|[^|]*\|([^|]*)/) {
			$account      = "\U$1";
			$user         = $2;
			$rawusage     = $3;

			if (exists($acc_limits{$account}) && $user eq $thisuser) {
				$user_usage{$account} = sprintf("%.0f", $rawusage/60);
			} elsif (exists($acc_limits{$account})) {
				$acc_usage{$account} = sprintf("%.0f", $rawusage/60);
			}
		}
	}

	close(SSHARE);


	# display formatted output
	printf "%-10s %11s | %16s %11s | %13s %11s (CPU hrs)\n",
		"User", "Usage", "Account", "Usage", "Account Limit", "Available";
	printf "%10s %11s + %16s %11s + %13s %11s\n",
		"-"x10, "-"x11, "-"x16, "-"x11, "-"x13, "-"x11;

	foreach my $acc (sort keys %user_usage) {
		printf "%-10s %11s | %16s %11s | %13s %11s\n",
			$thisuser, fmt_mins_as_hrs($user_usage{$acc}),
			$acc, fmt_mins_as_hrs($acc_usage{$acc}),
			fmt_mins_as_hrs($acc_limits{$acc}),
			fmt_mins_as_hrs($acc_limits{$acc} - $acc_usage{$acc});
	}
}



ikiwiki/relativedate.js
// Causes html elements in the 'relativedate' class to be displayed
// as relative dates. The date is parsed from the title attribute, or from
// the element content.

var dateElements;

hook("onload", getDates);

function getDates() {
	dateElements = getElementsByClass('relativedate');
	for (var i = 0; i < dateElements.length; i++) {
		var elt = dateElements[i];
		var title = elt.attributes.title;
		var d = new Date(title ? title.value : elt.innerHTML);
		if (! isNaN(d)) {
			dateElements[i].date=d;
			elt.title=elt.innerHTML;
		}
	}

	showDates();
}

function showDates() {
	for (var i = 0; i < dateElements.length; i++) {
		var elt = dateElements[i];
		var d = elt.date;
		if (! isNaN(d)) {
			elt.innerHTML=relativeDate(d);
		}
	}
	setTimeout(showDates,30000); // keep updating every 30s
}

var timeUnits = [
	{ unit: 'year',		seconds: 60 * 60 * 24 * 364 },
	{ unit: 'month',	seconds: 60 * 60 * 24 * 30 },
	{ unit: 'day',		seconds: 60 * 60 * 24 },
	{ unit: 'hour',		seconds: 60 * 60 },
	{ unit: 'minute',	seconds: 60 },
];

function relativeDate(date) {
	var now = new Date();
	var offset = date.getTime() - now.getTime();
	var seconds = Math.round(Math.abs(offset) / 1000);

	// hack to avoid reading just in the future if there is a minor
	// amount of clock slip
	if (offset >= 0 && seconds < 30 * 60 * 60) {
		return "just now";
	}

	var ret = "";
	var shown = 0;
	for (i = 0; i < timeUnits.length; i++) {
		if (seconds >= timeUnits[i].seconds) {
			var num = Math.floor(seconds / timeUnits[i].seconds);
			seconds -= num * timeUnits[i].seconds;
			if (ret)
				ret += "and ";
			ret += num + " " + timeUnits[i].unit + (num > 1 ? "s" : "") + " ";

			if (++shown == 2)
				break;
		}
		else if (shown)
			break;
	}

	if (! ret)
		ret = "less than a minute "

	return ret + (offset < 0 ? "ago" : "from now");
}
